### ISSUE TITLE
The DISPATCH() macro is not as efficient as it could be (move PyThreadState.use_tracing)

### ISSUE THREAD

**Issue by markshannon** [python/cpython#87926]
BPO | [43760](https://bugs.python.org/issue43760)
--- | :---
Nosy | @gvanrossum, @rhettinger, @scoder, @vstinner, @encukou, @ambv, @markshannon, @hroncok, @pablogsal, @miss-islington, @erlend-aasland, @Xtrem532
PRs | <li>python/cpython#25244</li><li>python/cpython#25276</li><li>python/cpython#28474</li><li>python/cpython#28498</li><li>python/cpython#28527</li><li>python/cpython#28529</li><li>python/cpython#28542</li><li>python/cpython#28723</li><li>python/cpython#29032</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-04-07.09:35:49.770>
labels = ['interpreter-core', 'expert-C-API', '3.10', 'performance']
title = 'The DISPATCH() macro is not as efficient as it could be (move PyThreadState.use_tracing)'
updated_at = <Date 2021-11-08.17:16:37.213>
user = 'https://github.com/markshannon'
```

bugs.python.org fields:
```python
activity = <Date 2021-11-08.17:16:37.213>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'C API']
creation = <Date 2021-04-07.09:35:49.770>
creator = 'Mark.Shannon'
dependencies = []
files = []
hgrepos = []
issue_num = 43760
keywords = ['patch']
message_count = 46.0
messages = ['390410', '390522', '390951', '393379', '393384', '393385', '393389', '393390', '393401', '393403', '393404', '393407', '393410', '393425', '393459', '393466', '393666', '402150', '402162', '402216', '402225', '402231', '402234', '402235', '402358', '402360', '402362', '402415', '402416', '402426', '402432', '402434', '402481', '402496', '402525', '402594', '402603', '403166', '403217', '404020', '404024', '404025', '404173', '404197', '404600', '405968']
nosy_count = 13.0
nosy_names = ['gvanrossum', 'rhettinger', 'jpe', 'scoder', 'vstinner', 'petr.viktorin', 'lukasz.langa', 'Mark.Shannon', 'hroncok', 'pablogsal', 'miss-islington', 'erlendaasland', 'Xtrem532']
pr_nums = ['25244', '25276', '28474', '28498', '28527', '28529', '28542', '28723', '29032']
priority = None
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'performance'
url = 'https://bugs.python.org/issue43760'
versions = ['Python 3.10']
```

</p></details>

**Comment by iritkatriel** [comment]
Is there anything left to do here?

**Comment by vstinner** [comment]
IMO we are done, and I close the issue. See the issue #84128 for making PyThreadState opaque.

>  The DISPATCH() macro is not as efficient as it could be

This part was fixed early. Following comments were more about PyThreadState incompatible changes, how to migrate existing C extensions to Python 3.11 and how to design a new API which no longer access directly PyThreadState changes. In fact, that's already the topic of the issue #84128 that I created in 2020 and so we can continue the discussion there.

The main change related to PyThreadState was the use_tracing member which was moved. I added PyThreadState_EnterTracing() and PyThreadState_LeaveTracing() functions to Python 3.11 for that and so projects already use it.

> I created https://github.com/python/cpython/pull/29121 to add PyThreadState_SetProfile() and PyThreadState_SetTrace() functions.

I abandoned this PR. @pablogsal added PyEval_SetProfileAllThreads() and PyEval_SetTraceAllThreads() functions to Python 3.12 (commit e34c82abeb7ace09e6b5d116585c47cc372996c1) which should fit the use case, with a different design.
